Utils/NamedTaskQueue.swift

```diff
@@ -0,0 +1,174 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+* License, v. 2.0. If a copy of the MPL was not distributed with this
+* file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+import Foundation
+import Deferred
+
+private let DefaultDispatchQueue = dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_DEFAULT, 0)
+
+// NamedTaskQueue is a mutable task queue, terminated by a `Deferred<Maybe<[String:T]>>`
+// The initiator task takes no arguments, and returns a Deferred<Maybe<S>>.
+// Once this has run, an appendable list of named tasks are run, of signature
+// `(S) -> Deferred<Maybe<T>>`.
+// You add named tasks, which are executed on a dispatch queue.
+// Once the tasks are all executed, then terminal is filled with the results from all tasks that were added.
+// If any task, including the initiator fails, then the terminal is filled with a failure Maybe.
+// This differs from `DeferredUtils.all` in that:
+//  * tasks can be added after construction.
+//  * only one task for a given name is executed.
+// Tasks are `(S) -> Deferred<Maybe<T>>`, added with the `append:name:task` method.
+// The result of running all the tasks are accessed either as a `Deferred<Maybe<[String:Maybe<T>]>>`, or by
+// adding an `upon` block to the queue.
+// If the queue has finished running all tasks, then no more tasks can be added, and a new queue is needed.
+public class NamedTaskQueue<S, T> {
+
+    public typealias Task = (S) -> Deferred<Maybe<T>>
+
+    public typealias NamedTask = (String, Task)
+
+    public typealias AccumulatedTaskResults = [String: Maybe<T>]
+
+    private let initialTask: () -> Deferred<Maybe<S>>
+
+    private let dispatchQueue: dispatch_queue_t
+
+    private let terminal: Deferred<Maybe<AccumulatedTaskResults>> = Deferred()
+
+    private let waitingTasks: LockProtected<[NamedTask]>
+
+    private let lock = NSRecursiveLock()
+
+    private var doneTasks: Set<String> = Set()
+
+    private var accumulator: AccumulatedTaskResults = [:]
+
+    private var isStarted: Bool = false
+
+    /**
+     * Has this task queue finished.
+     * Once the task queue has finished, it cannot have more tasks appended.
+     */
+    public var isFilled: Bool {
+        return terminal.isFilled
+    }
+
+    /**
+     * This is the deferred value that is the final result of the queue of tasks.
+     * It's given here so the mutable queue can plug in to other Deferreds.
+     */
+    public var deferred: Deferred<Maybe<AccumulatedTaskResults>> {
+        ensureStarted()
+        return terminal
+    }
+
+    public convenience init(inputValue: S, tasks: [NamedTask] = [], queue: dispatch_queue_t = DefaultDispatchQueue) {
+        let starter: () -> Deferred<Maybe<S>> = { deferMaybe(inputValue) }
+        self.init(initiator: starter, tasks: tasks, queue: queue)
+    }
+
+    public init(initiator: () -> Deferred<Maybe<S>>, tasks: [NamedTask] = [], queue: dispatch_queue_t = DefaultDispatchQueue) {
```

Comment:
To clarify: this is a system that behaves like an independent sequential reducer, so it needs an initial state `() -> S`, a sequence of tasks `S -> T`, and (implicitly) a post-accumulator `T, T, T -> [T]`.

You should address — in comment or code — why you're choosing to avoid the usual reducer pattern of `([T], (U, T) -> U) -> U`. In particular, the tasks you have here take an `S` as input, and it looks like that `S` never changes; it's passed to all subsequent tasks.

It's worth thinking about whether you're intending that `S` to be mutable shared state, or whether you're really intending there to be chaining, or if callers need to handle interdependence between tasks themselves.